### PR TITLE
En/review eg/losse val grad

### PR DESCRIPTION
**Pull request by eduardolneto**
Modifying losses.py to allow custom_losses to have a val_and_grad function instead of only a grad function. This is necessary for the augmented lagragian modifications in the parallel pull request, since in lbfgsb the val_and_grad option is more efficient and the only viable option. Changes on losses.py.

### CODE REVIEW

File: essos/losses.py [hunk 1]
```diff
@@ -71,23 +71,40 @@ def __init__(self, fun, *args_names, **kwargs):
...
+            self._starting_dofs, tuple_unraveler = ravel_pytree(
+                tuple(self.dependencies[arg] for arg in self.args_names)
+            )
+            self._dofs_to_args = tuple_unraveler
+
+            def _named_unraveler(dofs):
+                args = tuple_unraveler(dofs)
+                return {name: value for name, value in zip(self.args_names, args)}
+
+            self._dofs_to_pytree = _named_unraveler
```
Comment (EstevaoMGomes) [inline essos/losses.py:91]: This fixes the problem that we had, that the custom_loss did not have a named unraveler, only the composite loss, right?

Comment (EstevaoMGomes) [reply essos/losses.py:91]: I would need to test this to make sure everything works properly. In theory, it makes sense, but I need more time to figure it out properly.

File: essos/losses.py [hunk 2]
```diff
@@ -71,23 +71,40 @@ def __init__(self, fun, *args_names, **kwargs):
...
         self.args_names = args_names
         self.kwargs = kwargs
+        self._dofs_to_args = None
+
+    def clear_cache(self):
+        super().clear_cache()
+        self._dofs_to_args = None
+
+    def _ensure_unravelers(self):
+        if self._starting_dofs is None or self._dofs_to_args is None or self._dofs_to_pytree is None:
```
Comment (EstevaoMGomes) [inline essos/losses.py:81]: Maybe use `and` instead of `or`

**Review by EstevaoMGomes — COMMENTED** [review]
Great changes!

File: essos/losses.py (hunk 1 above)
Comment (eduardolneto) [reply essos/losses.py:91]: Yes, The examples to be added after were technically working. I guess we could add more later?

File: essos/losses.py (hunk 2 above)
Comment (eduardolneto) [reply essos/losses.py:81]: I am not sure I understand this change. If none of those are defined donn't we always want to define them?

**Review by rogeriojorge — CHANGES_REQUESTED** [review]
Mostly 3 minor stuff:

1. losses.py:111, custom_loss.call_pytree():
 dofs_to_pytree now returns a named dict, but call_pytree() still does self.fun(*dofs_pytree, **self.kwargs). If dofs_pytree is a dict, this passes the keys as positional args. I think this should either be:

args = tuple(dofs_pytree[name] for name in self.args_names)
return self.fun(*args, **self.kwargs)

or, if the custom loss functions are expected to accept keyword args:

return self.fun(**dofs_pytree, **self.kwargs)

2. losses.py:132-136, custom_loss.grad_pytree()

same things. It currently does *dofs_pytree, which will be wrong if dofs_pytree is the new named dict. I would build

args = tuple(dofs_pytree[name] for name in self.args_names)

and differentiate with those positional args, then put the gradients back into buffer.

3. losses.py:80, _ensure_unravelers

I think "or" is actually the right condition here. Using "and" would only rebuild when all of them are None, which could leave _dofs_to_args or _dofs_to_pytree unset. But not sure.

-----

A small test for custom_loss.value_and_grad() with two dependencies, and dofs_to_pytree() followed by call_pytree() / grad_pytree() would probably catch the dict-vs-tuple issue above.

File: essos/losses.py (hunk 2 above)
Comment (EstevaoMGomes) [reply essos/losses.py:81]: You are right.

File: essos/losses.py (hunk 1 above)
Comment (eduardolneto) [reply essos/losses.py:91]: Added some tests on test_multiobjectives.py

**Comment by eduardolneto** [comment]
I think I addressed all the comments for this PR.

**Comment by rogeriojorge** [comment]
Can you just add one assertion for value_and_grad() itself in test_custom_loss_named_unraveler()?
Otherwise, it's good for me to merge

**Review by rogeriojorge — APPROVED** [review]
Can you just add one assertion for value_and_grad() itself in test_custom_loss_named_unraveler()?
Otherwise, it's good for me to merge

**Comment by eduardolneto** [comment]
Added assertion for val_and_grad

File: tests/test_multiobjectives.py [hunk 3]
```diff
@@ -28,15 +29,50 @@ def mock_vmec():
...
-
-
-def test_build_available_inputs( vmec=mock_vmec(),  dummy_loss_fn=dummy_loss_fn()):
-    optimizer = MultiObjectiveOptimizer(
-        loss_functions=[dummy_loss_fn],
+def dummy_loss_fn():
+    def loss_fn(field=None, coils=None, vmec=None, surface=None, x=None):
+        return jnp.sum(x)
+    return loss_fn
+
```
Comment (EstevaoMGomes) [inline tests/test_multiobjectives.py:34]: This is terrible code. If you define just 
```
def loss_fn(field=None, coils=None, vmec=None, surface=None, x=None):
    return jnp.sum(x)
```
You can use loss_fn as a function without calling it. Also, with this, you always get an error when calling `dummy_loss_fn`, because you are doing `jnp.sum(None)`...

Comment (EstevaoMGomes) [reply tests/test_multiobjectives.py:34]: Also, if it is not used, it's better to just remove it.

**Review by EstevaoMGomes — COMMENTED** [review]
After fixing the test, this PR is ready to be merged.

File: tests/test_multiobjectives.py (hunk 3 above)
Comment (eduardolneto) [reply tests/test_multiobjectives.py:34]: @EstevaoMGomes  Do you actually understand that that part of the code is present on PR #29  and I did not add it here?

Comment (EstevaoMGomes) [reply tests/test_multiobjectives.py:34]: I think `tests/test_multiobjectives.py` was not altered in [PR!29](https://github.com/uwplasma/ESSOS/pull/29)...

Comment (eduardolneto) [reply tests/test_multiobjectives.py:34]: No, it was added on PR #21 with the multiobjectives.py, and it is on main. My point here is that you keep bringing things to be changed on these PR's that have nothing to do with the changes or what these PR's are addressing.

Comment (EstevaoMGomes) [reply tests/test_multiobjectives.py:34]: Sure, what im saying is just for them to be self-consistent. It’s fine either way

Comment (eduardolneto) [reply tests/test_multiobjectives.py:34]: Solved unnecessary wrapper.
